lib/grpc_provider.cpp

```diff
@@ -0,0 +1,284 @@
+                        game, player, msg.seethrough());
+                case UserOperation::kSwapCards:
+                    return Transition<ExplodingKittensProto::SwapCards>(
+                        game, player, msg.swapcards());
+                case UserOperation::kGetBottom:
+                    return Transition<ExplodingKittensProto::GetBottom>(
+                        game, player, msg.getbottom());
+                case UserOperation::kShuffle:
+                    return Transition<ExplodingKittensProto::Shuffle>(
+                        game, player, msg.shuffle());
+                case UserOperation::kExtort:
+                    return Transition<ExplodingKittensProto::Extort>(
+                        game, player, msg.extort());
+                case UserOperation::kExtortedRespond:
+                    return Transition<ExplodingKittensProto::ExtortedRespond>(
+                        game, player, msg.extortedrespond());
+                case UserOperation::kBombDisposal:
+                    return Transition<ExplodingKittensProto::BombDisposal>(
+                        game, player, msg.bombdisposal());
+                case UserOperation::kDrawCard:
+                    return Transition<ExplodingKittensProto::DrawCard>(
+                        game, player, msg.drawcard());
+            }
+        }
+    }
+}
+
+void Provider::SendNotifyMsg(
+    int roomid, int uid, const ExplodingKittensProto::NotifyMsg& notifyMsg) {
+    Core::ProviderMsg providerMsg;
+    auto providerArgs = providerMsg.mutable_notifymsgargs();
+    providerArgs->set_roomid(roomid);
+    providerArgs->set_userid(uid);
+    providerArgs->mutable_custom()->PackFrom(notifyMsg);
+
+    mStream->Write(providerMsg);
+}
+
+void Provider::SendCardOperationRespond(int roomid, int uid,
+                                        ExplodingKittensProto::CardType card,
+                                        const std::vector<int>& uids,
+                                        int targetuid) {
+    ExplodingKittensProto::NotifyMsg notifyOperRespondMsg;
+    ExplodingKittensProto::CardOperation* cardOperation =
+        notifyOperRespondMsg.mutable_cardoperation();
+    cardOperation->set_userid(uid);
+    cardOperation->set_card(card);
+    cardOperation->set_targetuid(targetuid);
+    cardOperation->set_isdrawcard(card == ExplodingKittensProto::NONE ? true
+                                                                      : false);
+    for (auto id : uids) {
+        SendNotifyMsg(roomid, id, notifyOperRespondMsg);
+    }
+}
+
+void Provider::SendSwapResult(int roomid, int uid,
+                              const std::vector<CardType>& cards) {
+    ExplodingKittensProto::NotifyMsg notifySwapRetMsg;
+    ExplodingKittensProto::SwapResult* swapResult =
+        notifySwapRetMsg.mutable_swapresult();
+    *(swapResult->mutable_cards()) =
+        google::protobuf::RepeatedField<google::protobuf::int32>(cards.begin(),
```

Comment:
As the long definition occus again and again.
Could you write a template func?
1. 
template<typename MsgT>
inline void PopulateCards(MsgT* msg, const std::vector<CardType>& cards)
{
*(msg->mutable_cards()) = google::protobuf::RepeatedField<google::protobuf::int32>(cards.begin(), cards.end());
}

there is one ugly part: requires the method name is mutable_cards.
I guess most of the use case can live with that.
but,  if you really have different name.

I would propose:
2.
inline void PopulateCards(google::protobuf::RepeatedField<google::protobuf::int32>* grpcCards, const std::vector<CardType>& cards)
{
*grpcCards= google::protobuf::RepeatedField<google::protobuf::int32>(cards.begin(), cards.end());
}

Reply:
done